### PR TITLE
Add VectorOfVariables-in-SecondOrderCone

### PR DESCRIPTION
**Pull request by odow**
There are a few todos:

In Gurobi.jl

- [x] Figure out how to set `t >= 0`

In MathOptInterface.jl

- [x] <s>Fix soc3 test to allow `INFEASIBLE_OR_UNBOUNDED` as a return https://github.com/JuliaOpt/MathOptInterface.jl/issues/836 https://github.com/JuliaOpt/MathOptInterface.jl/pull/839</s> We set `DualReductions=0`.
- [x] Implement `ConstraintFunction` and `ConstraintSet` in `geomean` bridge https://github.com/JuliaOpt/MathOptInterface.jl/issues/750

image

### CODE REVIEW

File: src/MOI_wrapper.jl [hunk 1]
```diff
@@ -2525,3 +2554,140 @@ function MOI.supports(
...
+        error("Dimension of $(s) does not match number of terms in $(f)")
+    end
+    # SOC is the cone: t ≥ ||x||₂ ≥ 0. In quadratic form, this is
+    # t² - Σᵢ xᵢ² ≥ 0 and t ≥ 0.
+
+    # First, check the lower bound on t.
+
+    # TODO(odow): a better way to set the lower bound.
+    t_info = _info(model, f.variables[1])
+    set_dblattrelement!(model.inner, "LB", t_info.column, 0.0)
```
Comment (odow) [inline src/MOI_wrapper.jl]: Any ideas for how to do this? Because what if `t` already has a lower bound? I almost want to add a dummy variable with `t = t_dummy`, but then I'd have to account for the un-used variable elsewhere in the wrapper, leading to more complexity.

Comment (joaquimg) [reply src/MOI_wrapper.jl]: The MOI wrapper would know if there is an existing LB.
If there is one I would check what it is, if its less than zero, maybe, throw a warning?
If there is no LB we could add a the LB constraint OR only add this in the solver level and have an extra field in the optimizer to take car of this soc lb list of variables.

File: src/MOI_wrapper.jl [hunk 2]
```diff
@@ -230,9 +235,26 @@ function MOI.supports_constraint(
     return true
 end
 
+function MOI.supports(
+    model, ::Optimizer, ::Union{MOI.ConstraintSet, MOI.ConstraintFunction},
```
Comment (blegat) [inline src/MOI_wrapper.jl]: That shouldn't be defined:
```julia
julia> using MathOptInterface

julia> const MOI = MathOptInterface
MathOptInterface

julia> MOI.is_copyable(MOI.ConstraintSet())
false
```

File: src/MOI_wrapper.jl (hunk 1 above)
Comment (blegat) [reply src/MOI_wrapper.jl]:
```julia
if !MOI.is_valid(model, MOI.ConstraintIndex{MOI.SingleVariable, MOI.GreaterThan{Float64}(f.variables[1].value))
    # check that the bound is nonnegative, otherwise do
    MOI.add_constraint(model, MOI.SingleVariable(f.variables[1]), MOI.GreaterThan(0.0)) # This constraint should be hidden because it has not been added by the user. This is what `src/Bridges/bridge_optimizer.jl` does
end
```

File: src/MOI_wrapper.jl (hunk 2 above)
Comment (odow) [reply src/MOI_wrapper.jl]: What's the rule for needing to `support` attributes? If `is_copyable` is `true`?

Comment (blegat) [reply src/MOI_wrapper.jl]: Yes :)
https://github.com/JuliaOpt/MathOptInterface.jl/blob/0ed068ff8342cf2501305963c22708b0b18a4b80/src/attributes.jl#L157-L159

**Comment by odow** [comment]
I have a slightly ugly implementation that passes tests (of which I've added a lot to catch different edge cases. They should go into MOI). Opinions?

**Comment by odow** [comment]
image

**Comment by odow** [comment]
image

File: src/MOI_wrapper.jl [hunk 3]
```diff
@@ -877,14 +883,70 @@ function MOI.delete(
...
+        set_dblattrelement!(model.inner, "LB", info.column, value)
+        _require_update(model)
+    elseif value >= 0.0
+        # Regardless of whether there are SOC constraints, this is a valid bound
+        # for the SOC constraint and should over-ride any previous bounds.
+        info.lower_bound_if_soc = NaN
+        set_dblattrelement!(model.inner, "LB", info.column, value)
+        _require_update(model)
+    elseif isnan(info.lower_bound_if_soc)
+        # Previously, we had a +ve lower bound (i.e., it was set in the case
```
Comment (mlubin) [inline src/MOI_wrapper.jl]: It's better to write out positive and negative. It took a few seconds for me to figure out what +ve and -ve meant.
